### PR TITLE
Added path tracing renderer

### PR DESCRIPTION
**Pull request by favreau**

### CODE REVIEW

File: brayns/tasks/LoadDataFunctor.cpp [hunk 1]
```diff
@@ -49,11 +49,11 @@ void LoadDataFunctor::operator()(Blob&& blob)
     {
         struct Scope
         {
-            Scope(Blob&& blob, LoadDataFunctor& parent)
+            Scope(Blob&& b, LoadDataFunctor& parent)
```
Comment (tribal-tec) [inline brayns/tasks/LoadDataFunctor.cpp]: why this change?

Comment (favreau) [reply brayns/tasks/LoadDataFunctor.cpp]: Clang complains about variable shadowing the other one line 62. Gcc is happy though. I had this issue when compiling on MacOSX

Comment (tribal-tec) [reply brayns/tasks/LoadDataFunctor.cpp]: ah ok

File: plugins/engines/ospray/ispc/render/PathTracingRenderer.ispc [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+
+        // compute ray direction of uniformly random diffuse ray
+        const vec3f raydir =
+            getRandomVector(sample, dg.Ns, self->abstract.super.epsilon);
+
+        // origin of new ray in path is hitpoint of previous ray in path
+        const vec3f hitpoint = dg.P + self->abstract.super.epsilon * dg.Ns;
+        setRay(localray, hitpoint, raydir);
+        localray.t0 = self->abstract.super.epsilon;
+        localray.t = self->abstract.ambientOcclusionDistance;
```
Comment (straaljager) [inline plugins/engines/ospray/ispc/render/PathTracingRenderer.ispc]: I'm not sure if limiting t (the maximum allowed hit distance along ray) to the ambientOcclusionDistance will work the same as setting it to a very large number. This may or may not result in some weird lighting, but may also be faster.

File: plugins/engines/ospray/ispc/render/PathTracingRenderer.ispc [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+                        // create and trace a shadowray towards the sun
+                        Ray sunRay;
+                        setRay(sunRay, hitpoint, sunSampleDirection);
+                        if (isOccluded(self->abstract.super.model, sunRay))
+                            // take surface color at this bounce's hitpoint into
+                            // account
+                            Kd = Kd * (1.f - self->abstract.shadows);
+                    }
+                    else
+                        Kd = Kd * (1.f - self->abstract.shadows);
```
Comment (straaljager) [inline plugins/engines/ospray/ispc/render/PathTracingRenderer.ispc]: This doesn't seem right, the if/else statement has the same outcome for both the if and else branch (if the surface faces the light source and is not occluded from it, then the hitpoint should receive additional light).

File: plugins/engines/ospray/ispc/render/PathTracingRenderer.ispc [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+                const varying Light_SampleRes lightSample =
+                    light->sample(light, dg, s);
+                const vec3f radiance = lightSample.weight;
+                if (reduce_max(radiance) > 0.f)
+                {
+                    const vec3f lightDirection = lightSample.dir;
+                    const vec3f sunSampleDirection =
+                        getConeSample(lightDirection, rng,
+                                      self->abstract.softShadows);
+                    const float sunLight =
```
Comment (straaljager) [inline plugins/engines/ospray/ispc/render/PathTracingRenderer.ispc]: I would rename sunLight to sunAngle to avoid confusion and reflect better what it's doing (that was a bad naming choice on my part, sorry)

File: plugins/engines/ospray/ispc/render/PathTracingRenderer.ispc [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+                        Kd = Kd * (1.f - self->abstract.shadows);
+                }
+            }
+        }
+        mask = mask * Kd;
+    } // end bounces
+
+    // CHECK THIS!!! WITH OR WITHOUT * MASK
+    // results are very different. Color bleeding is much stronger with mask
+    color = accucolor * mask;
```
Comment (straaljager) [inline plugins/engines/ospray/ispc/render/PathTracingRenderer.ispc]: After some digging into path tracing theory, it would be more correct to not multiply the final color by the mask colour again, as it has already been taken into account. Otherwise this will result into too much colour bleeding.

File: plugins/engines/ospray/ispc/render/PathTracingRenderer.ispc (hunk 2 above)
Comment (favreau) [reply plugins/engines/ospray/ispc/render/PathTracingRenderer.ispc]: I understand, but the ambient occlusion distance can be set to a very large value anyway. And this way, we can change it at runtime.

File: plugins/engines/ospray/ispc/render/PathTracingRenderer.ispc (hunk 3 above)
Comment (favreau) [reply plugins/engines/ospray/ispc/render/PathTracingRenderer.ispc]: Hum, you're right, I am not 100% sure either of the current implementation. Let's talk tomorrow.

File: plugins/engines/ospray/ispc/render/PathTracingRenderer.ispc (hunk 4 above)
Comment (favreau) [reply plugins/engines/ospray/ispc/render/PathTracingRenderer.ispc]: Sure, I usually use cosNL for diffuse and cosLR for specular. In this case, only diffuse shading is processed. And Sun is probably a specific case too since we can have many lights. I'll change it to cosNL then, unless you have an objection.

File: plugins/engines/ospray/ispc/render/PathTracingRenderer.ispc (hunk 5 above)
Comment (favreau) [reply plugins/engines/ospray/ispc/render/PathTracingRenderer.ispc]: Ok. Fixed! :)
